File: MoliStudy/AccountDao.m
//
//  AccountDao.m
//  MoliStudy
//
//  Created by zhaoqin on 11/22/15.
//  Copyright © 2015 MoliStudy. All rights reserved.
//

#import "AccountDao.h"
#import "Account.h"
#import "AppDelegate.h"

@implementation AccountDao{
    AppDelegate *appDelegate;
    NSManagedObjectContext *appContext;
}

static AccountDao* sharedManager = nil;

- (id)init{
    self = [super init];
    appDelegate = [[UIApplication sharedApplication] delegate];
    appContext = [appDelegate managedObjectContext];
    return self;
}

- (void)save{
    [appDelegate saveContext];
}

+ (AccountDao*)sharedManager{
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        sharedManager = [[self alloc] init];
        NSEntityDescription *entityDescription = [NSEntityDescription entityForName:@"Account" inManagedObjectContext:sharedManager->appContext];
        NSFetchRequest *request = [[NSFetchRequest alloc] init];
        [request setEntity:entityDescription];
        NSError *error = nil;
        NSArray *array = [sharedManager->appContext executeFetchRequest:request error:&error];
        if([array count] != 0){
            sharedManager.account = array[0];
        }
    });
    
    return sharedManager;
}

- (void)add:(Account *)model{
    self.account = model;
    [self save];
}

- (void)modified:(Account *)model{
    self.account.accountName = model.accountName;
    self.account.password = model.password;
    self.account.userID = model.userID;
    self.account.userName = model.userName;
    self.account.currentAcademy = model.currentAcademy;
    self.account.targetAcademy = model.targetAcademy;
    [self save];
}

- (Account*)findAccount{
    return self.account;
}

@end

File: MoliStudy/AccountDAO.swift
//
//  AccountDAO.swift
//  MoliStudy
//
//  Created by zhaoqin on 11/30/15.
//  Copyright © 2015 MoliStudy. All rights reserved.
//

import UIKit

class AccountDAO: NSObject {
    
    static let sharedManager = AccountDAO()
    var isExist = false
    
    var account: Account!
    
    let managedContext = ContextDAO.sharedInstance.appContext
    
    override init() {
        //Create a net fetch request using the Account entity
        let fetchRequest = NSFetchRequest(entityName: "Account")
        
        do{
            
            let results = try managedContext.executeFetchRequest(fetchRequest)

            if results.count != 0{
                self.account = results[0] as! Account
                isExist = true
            }
            print(results.count)
            print(isExist)
        }catch let error as NSError{
            print("Could not fetch \(error), \(error.userInfo)")
            
        }
    }
    
    func addAccount(accountName: String, password: String, userID: String, userToken: String){
        if !isExist{
            print("isExist")
            account = NSEntityDescription.insertNewObjectForEntityForName("Account", inManagedObjectContext: managedContext) as! Account
        }
        account.accountName = accountName
        account.password = password
        account.userID = userID
        account.token = userToken
        save()
    }
    
    func completeInfo(userName: String, sex: NSNumber, academy: String, qq: String, introduction: String){
        account.userName = userName
        account.sex = sex
        account.academy = academy
        account.qq = qq
        account.introduction = introduction
        save()
    }
    
    func save(){
        ContextDAO.sharedInstance.appDelegate.saveContext()
    }
    
    func findAccount() -> Account{
        return self.account
    }
}
